crates/binjs_generate_library/src/lib.rs

```diff
@@ -429,13 +386,49 @@ impl<'a> Walker<'a> for {name} where Self: 'a {{
                     let definition = format!(
                         "
 /// Implementation of interface sum {node_name}
-#[derive(PartialEq, Debug, Clone)]
+#[derive(PartialEq, Debug, Clone, Serialize)]
+#[serde(tag = \"type\")]
 pub enum {name} {{
 {contents}
     /// An additional value used to mark that the node was stolen by a call to `steal()`.
     BinASTStolen,
 }}\n
 
+// An optimised implementation of tagged deserialise that expects `type` to be the first key in the object.
```

Comment:
Won't that make the deserializer incompatible with many tools and possibly with future versions of Shift?

How difficult would it be to implement a fallback that goes the slow way? Not necessarily in this patch, but maybe in a followup.

Reply:
Yeah, in theory it could, but:

1. We currently have only one well-defined source from which we currently deserialise these ASTs.
2. In practice I haven't seen real-world cases where tagged enums wouldn't have this `type` field as first in the object, especially if it comes from well-defined typed structures (like Shift AST or Rust `derive(Serialize)`).
3. Implementation raises a clear error if this contract is violated.

If someday we deserialise AST from a source where `type` field does happen not to be the first one in the JSON, we'll notice a clear error from (3) right away and might fix by either going back to `derive(Deserialize)` or modifying the transform on the JS side or buffering into an intermediate AST ourselves.

Until then, given how bad this slow path is, it seems better to raise an error than try to silently accept it.

Btw, later turned out that comment might be not 100% correct - I've been pointed out that JSON spec does allow implementors to assign semantic meaning to the order of keys, so we don't violate it. https://twitter.com/awbjs/status/1106287165892288512

Reply:
Works for me.